### ISSUE TITLE
TypeError: Cannot read property 'indexOf' of undefined

### ISSUE THREAD

**Issue by PDiddyDaddyDog** [jvmahon/homebridge-homeseer#78]
Howdy,

I apologize in advance if this has been solved in a previous issue post, I did look but I'm afraid I am but a noob and couldn't figure it out.
This is a Windows 10 setup of Homebridge with the Homeseer Plug-in.
Everything was going well after the install, the QR code came up and HomeKit could pair with it.
When creating the config.json file I copied the content of the samples provided without much success and eventually came up the file attached. (.txt extension )
config.json.txt

Anyhoo, please help and I apologize again if this is a duplicate of a previously solved issue.

Log output
C:\WINDOWS\system32>homebridge
[2019-4-16 18:37:54] Loaded config.json with 4 accessories and 1 platforms.
[2019-4-16 18:37:54] ---
[2019-4-16 18:37:56] Loaded plugin: homebridge-homeseer-plugin-2018
homebridge API version: 2.4
[2019-4-16 18:37:56] Registering platform 'homebridge-HomeSeerPlatform.HomeSeer'
[2019-4-16 18:37:56] ---
[2019-4-16 18:37:56] Loading 1 platforms...
[2019-4-16 18:37:56] [HomeSeer] Initializing HomeSeer platform...
[2019-4-16 18:37:56] [HomeSeer] System default periodic polling rate set to: 10 seconds
[2019-4-16 18:37:56] Loading 4 accessories...
[2019-4-16 18:37:56] TypeError: Cannot read property 'indexOf' of undefined
at API.accessory (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\lib\api.js:50:12)
at Server._loadAccessories (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\lib\server.js:297:42)
at Server.run (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\lib\server.js:91:38)
at module.exports (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\lib\cli.js:59:10)
at Object. (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\bin\homebridge:17:22)
at Module._compile (module.js:653:30)
at Object.Module._extensions..js (module.js:664:10)
at Module.load (module.js:566:32)
at tryModuleLoad (module.js:506:12)
at Function.Module._load (module.js:498:3)

C:\WINDOWS\system32>
[config.json.txt](https://github.com/jvmahon/homebridge-homeseer/files/3088371/config.json.txt)

**Comment by jvmahon** [comment]
I was not able to duplicate the error and its pretty odd since youre JSON file is so simple. But I've simplified it a bit more. Try the attached and let me know if it works.

[config.json.txt](https://github.com/jvmahon/homebridge-homeseer/files/3088821/config.json.txt)

The few things I can think of . . .

1. Are you running the latest version of HomeSeer (version 3.0.0.531) - I recall one of the more recent updates fixed something in the JSON interface. If you're not running the most recent HomeSeer, update it from here: [HomeSeer Downloads](https://homeseer.com/current-downloads/)

2. Is HomeSeer running on the same computer as HomeBridge?  If not, you need to change the "Hosts" 127.0.0.1 IP address to the IP address of the computer with HomeSeer

3. Anything "unusual" in the HomeSeer device names?  Check the names in homeseer - just use standard ASCII letters and spaces - no punctuation.  (I have no idea if this could be the issue, but its something to try).

4. You said you had multiple failures trying to set up the config.json.  Maybe there was something "left behind" that's causing the failures. You'll need to clean up your .homebridge folder. That's the folder where you placed the config.json. Delete the "accessories" and "persist" sub folders and try pairing again.

5. Install the latest nodejs. [NodeJS Download](https://nodejs.org/en/)

I'm just making guesses here as I can't replicate the problem, Let me know if any of this helps.

**Comment by henlagt** [comment]
I don't know if my problem is related to the same, I received a similar error message when I try to add  new devices to my config.json. I have gone through the news device and can't find anything wrong with them.

The error:
[HomeSeer] Checking Configuration Data
[HomeSeer] Cannot read property 'length' of undefined

My [config.json](https://github.com/henlagt/homeseer/blob/master/config.json)
One example is when I added this after line 105:
{"ref": 675, "type": "LightSensor", "name":"Watts TV Stue","uuid_base": "Lux675"},

HomeSeer = 3.0.0.531
homebridge = 0.0.48
homebridge-homeseer-plugin-2018 = 2.3.19

**Comment by jvmahon** [comment]
@henlagt. My best guess is that these are related problems, but I don't have enough information to fix i
Is this on Windows or Linux?
If it is on Windows, which version of Windows?
Which version of node are you using (from a command line, execute "node -v")?
If on windows, will you allow me to connect to your computer using Windows Remote assistance  (the "msra" command) so I can observe the error as it happens? If so, we can try and work out a time to do this.
Can you post more of the screen out put (right before the error happens)?

**Comment by henlagt** [comment]
This is Linux (raspbian strecht): raspberrypi 4.14.98-v7+

node -v = v11.14.0

More of the output:

Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] Loading 2 platforms...
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] [Config] Initializing config platform...
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] [Config] Spawning homebridge-config-ui-x with PID 4497
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] [HomeSeer] Initializing HomeSeer platform...
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] [HomeSeer] System default periodic polling rate set to: 10 seconds
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] Loading 0 accessories...
Apr 21 09:46:27 raspberrypi homebridge[4473]: Load homebridge-HomeSeerPlatform.HomeSeer
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] [HomeSeer] Checking Configuration Data
Apr 21 09:46:27 raspberrypi homebridge[4473]: [4/21/2019, 9:46:27 AM] [HomeSeer] Cannot read property 'length' of undefined
Apr 21 09:47:55 raspberrypi systemd[1]: Started Node.js HomeKit Server.

**Comment by jvmahon** [comment]
@henlagt. Can you send me a copy of your HomeSeer "Backup Config" file (created from the HomeSeer [Tools] -> [Setup] menu?  I need to try and re-create your setup to see if I can reproduce the error. You can send it to my email at jvm33@columbia.edu

**Comment by jvmahon** [comment]
@henlagt. In your config.json file, can you please try deleting the line
`````
accessories =[]
`````

which is on line 115 of your config.json
also delete the comma immediately before this (on line 114).

I think the code does not like the empty "accessories" array and it is not needed.

Let me know if this works. I'll then put in more "permanent" code into the plugin to guard against this error in the future.

**Comment by henlagt** [comment]
I removed it, but it didn't change the error when adding a new accessories. I removed it from my working config.json so it don't make any problems later.

**Comment by jvmahon** [comment]
OK. 
I'll investigate this and let you know if I can come up with a solution.

**Comment by jvmahon** [comment]
@henlagt.  I have found the error. Homekit is limited to about 100 devices on a bridge. When you add your device 675, you end up with 1 too many devices which is supposed to display a warning. But there is a bug in that warning code that causes a crash. I'll fix that.  But more importantly, you need to split your devices into two confif.json files and then run two instances of homebridge so you don't exceed 100 devices on the bridge. There are instructions in the Wiki about this. 

See this link: https://github.com/jvmahon/homebridge-homeseer/wiki/Running-Multiple-Instances-of-HomeBridge

**Comment by jvmahon** [comment]
@henlagt. I've fixed the warning code and published a new version of the plugin. You need to do an update to the plugin.
`````
npm -g update
`````

You will still need to split your config.json file into smaller sections as explained in my prior post.

**Comment by henlagt** [comment]
I added a second instance a splittet my config.json, it works great.
Thank you for your help, really appreciate it.

**Comment by jvmahon** [comment]
@henlagt. By the way, I'm curious - it looks like some of your devices are Philips Hue devices (JowiHue). The Philips bridge (at least since Bridge version 2.0) can be directly paired to HomeKit -- why bother going through HomeSeer for those devices? Wouldn't it be better to use the Philips bridge for those devices, and use my plugin only for the remainder?

**Comment by rrman** [comment]
I also use Homebridge-HomeSeer-2019 as well as the JowieHue plug-in (for my Philips Hue lights). The Philips Hue bridge can talk directly to HomeKit so there is no need to go through the Homebridge-HomeSeer-2018 plug-in for the Philips lights. For everything else (e.g.  my Z-Wave devices), I use the Homebridge-HomeSeer-2018 plug-in.

**Comment by henlagt** [comment]
Many of my HUE devices are connected to HomeSeer through [raspbee](https://phoscon.de/en/raspbee). I'm consider to move the rest of the bulbs to raspbee.
I use HomeSeer with motion and lux sensors to automate all lights and the Home-app is used most as a backup and to improve the WAF :)

**Comment by PDiddyDaddyDog** [comment]
Hi Again,
Just released I had installed the 32 bit version of OpenSSL so have installed Win64OpenSSL-1_0_2r.exe as Win64OpenSSL-1_0_2q.exe was not available.
Also I have update homeseer to the lastest version as suggesed
Now when I try to run Homebridge I get the following.

openssl config failed: error:02001003:system library:fopen:No such process
C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\node_modules\bindings\bindings.js:91
throw e
^

Error: \?\C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\node_modules\ed25519-hap\build\Release\ed25519.node is not a valid Win32 application.
\?\C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\node_modules\ed25519-hap\build\Release\ed25519.node
at Object.Module._extensions..node (internal/modules/cjs/loader.js:857:18)
at Module.load (internal/modules/cjs/loader.js:685:32)
at Function.Module._load (internal/modules/cjs/loader.js:620:12)
at Module.require (internal/modules/cjs/loader.js:723:19)
at require (internal/modules/cjs/helpers.js:14:16)
at bindings (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\node_modules\bindings\bindings.js:84:48)
at Object. (C:\Users\andre\AppData\Roaming\npm\node_modules\homebridge\node_modules\ed25519-hap\index.js:1:37)
at Module._compile (internal/modules/cjs/loader.js:816:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:827:10)
at Module.load (internal/modules/cjs/loader.js:685:32)

**Comment by jvmahon** [comment]
O.K.,  the "r" version of OpenSSL should be fine.

You are in a zone of errors I haven't encountered before, but I suspect that what has happened is when you installed the npm modules, they tried to link with the 32 bit version of openSSL, and now there is an error since some of them may still be looking for that. I suggest you uninstall the npm modules and re-install them. Let me know if that works.

`````
npm -g uninstall hap-nodejs homebridge homebridge-homeseer-plugin-2018
npm -g install hap-nodejs homebridge homebridge-homeseer-plugin-2018
`````

The only other suggestion I have is I can try to help you using Windows Quick Assist, but to do that, you should contact me on my email at jvm33@columbia.edu.

**Comment by PDiddyDaddyDog** [comment]
So, after uninstalling and reinstalling the npm node modules, updating the config.json with your edits and cleaning up the homebridge folder, I'm back to the original problem.  Progress of sorts I guess.

**Comment by jvmahon** [comment]
@PDiddyDaddyDog 
In order to help you, I need to re-create your HomeSeer configuration. Can you do a backup of your HomeSeer setup (from the [Tools] -> [Setup] -> [General] tab) and send it to me at the email address I gave above (putting it on Dropbox or other sharing site also works). Please also include your config.json.

**Comment by jvmahon** [comment]
@PDiddyDaddyDog. I think I found your error. The config.json file was set up incorrectly, but it was very heard to see the error. The "accessories" group needs to be located within the Homeseer platform block.  specifically, it immediately follows the line: 
"ASCIIport":11000,

Try the attached config.json file.

[config.json.PDiddy.txt](https://github.com/jvmahon/homebridge-homeseer/files/3116441/config.json.PDiddy.txt)


Also, as an FYI - there has been an update to the hap-nodejs code which is what homebridge is based on. With this update, OpenSSL is no longer needed. You can install the update using:
`````
npm -g update
`````

**Comment by jvmahon** [comment]
@PDiddyDaddyDog .  Did the last post help you get it working? If so, then I'll close this comment out.

**Comment by PDiddyDaddyDog** [comment]
Hi again, sorry for the delay in replying. I've been painting the house and fell off a ladder and been kind of out of it for a few days.
Anyway, applied the updated config.json file and ran the npm update.
This is what I now see when trying to run Homebridge.

C:\WINDOWS\system32>homebridge
[4/29/2019, 11:07:17 PM] Loaded config.json with 0 accessories and 1 platforms.
[4/29/2019, 11:07:17 PM] ---
[4/29/2019, 11:07:18 PM] Loaded plugin: homebridge-homeseer-plugin-2018
homebridge API version: 2.4
[4/29/2019, 11:07:18 PM] Registering platform 'homebridge-HomeSeerPlatform.HomeSeer'
[4/29/2019, 11:07:18 PM] ---
[4/29/2019, 11:07:18 PM] Loading 1 platforms...
[4/29/2019, 11:07:18 PM] [HomeSeer] Initializing HomeSeer platform...
[4/29/2019, 11:07:18 PM] [HomeSeer] System default periodic polling rate set to: 10 seconds
Load homebridge-HomeSeerPlatform.HomeSeer
[4/29/2019, 11:07:19 PM] [HomeSeer] undefined

C:\WINDOWS\system32>

I think we are getting closer to a fix. :)

**Comment by jvmahon** [comment]
Hope you're feeling better.
Can you contact me on my email jvm33@columbia.edu

Just to be sure - are Homebridge and HomeSeer running on the same computer? And you started up HomeSeer before starting up HomeBridge?
